packages/database/src/core/PersistentConnection.ts

```diff
@@ -948,11 +948,11 @@ export class PersistentConnection extends ServerActions {
 
     // Puts depend on having received the corresponding data update from the server before they complete, so we must
     // make sure to send listens before puts.
-    forEach(this.listens_, (_pathString, queries) => {
-      forEach(queries, (_key, listenSpec) => {
+    for (const queries of Object.values(this.listens_)) {
```

Comment:
Object.values() already loops through the object once (see https://github.com/es-shims/Object.values/blob/master/implementation.js).

While not the most beautiful JS to write in 2019, this would be more efficient as:

```
for (const key in this.listens_) {
  if (this.listens_. hasOwnProperty(key)) {
     ...
  }
}
```

This applies in other places too. Sorry for not pointing this out earlier (or rather: Sorry for leading you down this path).

Reply:
Used a `Map` here instead of a plain `object`.

Unless these objects have 1000s of values, the performance difference is negligible and I'd rather have more readable code, but it's your package, so up to you.